doc/tomb-status.1
.TH tomb 1 "Sept 26, 2011" "tomb"

.SH NAME
Tomb \- the Crypto Undertaker

.SH SYNOPSIS
.B
.IP "tomb [options] command [arguments]"
.B
.IP "tomb-open [file]"
.B
.IP "tomb-status mountpoint"

.SH DESCRIPTION

Tomb is an application to manage the creation and access of encrypted
storage files: it can be operated from commandline and it can
integrate with a user's graphical desktop.

Tomb generates encrypted storage files to be opened and closed using
their associated keys, which are also protected with a password chosen
by the user. To create, open and close tombs a user will need super
user rights to execute the tomb commandline utility.

A tomb is like a locked folder that can be safely transported and
hidden in a filesystem; it encourages users to keep their keys
separate from tombs, for instance keeping a tomb file on your computer
harddisk and its key file on a USB stick.

For simplified use, the command \fItomb-open\fR starts a wizard that
guides users in the creation of a new tomb or, if a tomb file is
specified as \fIargument\fR, it opens it and makes it accessible in a
default location under the /media folder, starting the status tray
applet (\fItomb-status\fR) if a desktop is present.


.SH COMMANDS

.B
.IP "create"
Creates a new encrypted storage tomb and its key, named as specified
by the given \fIargument\fR.

.B
.IP "open"
Opens an existing tomb file specified in the \fIfirst argument\fR. If
a \fIsecond argument\fR is given it will indicate the \fImountpoint\fR
where the tomb should be made accessible, if not then the tomb is
mounted in a directory named after the filename and inside /media.

.B
.IP "list"

List all the tombs found open, including information about the time
they were opened and the hooks that they mounted. If the \fIfirst
argument\fR is present, then shows only the tomb named that way or
returns an error if its not found.

.B
.IP "close"
Closes a currently open tomb.  When \fIan argument\fR is specified, it
should be the name of a mounted tomb; if not specified and only one
tomb is open then it will be closed; if multiple tombs are open, the
command will list them on the terminal. The special
\fIargument\fR 'all' will close all currently open tombs. This command
fails if the tomb is in use by running processes, the command
\fIslam\fR can be used to force close.

.B
.IP "slam"
Closes a tomb like the command \fIclose\fR does, but in case it is in
use looks for all the processes accessing its files and violently
kills them using \-9.

.B
.IP "bury"
Hides a tomb key (\fIfirst argument\fR) inside a jpeg image (\fIsecond
argument\fR) using steganography: the image will change in a way that
cannot be noticed by human eyes and the presence of the key inside it
isn't detectable without the right password. This option is useful to
backup tomb keys in unsuspected places; it uses steghide and the
serpent encryption algorithm.

.B
.IP "exhume"
Extracts a named tomb key (\fIfirst argument\fR) from a (jpeg) image file
(\fIsecond argument\fR) known to be containing it, if the right password is
given. This is used to recoved buried keys from unsuspected places.

.SH OPTIONS
.B
.B
.IP "-s \fI<MBytes>\fR" 
When creating a tomb, this option  MUST be used to specify the size of
the new \fIfile\fR to be created, in megabytes.
.B
.IP "-k \fI<keyfile>\fR"
When opening a  tomb, this option can be used  to specify the location
of the  key to use. Keys  are created with  the same name of  the tomb
file adding a '.gpg' suffix,  but can be later renamed and transported
on other media. When a key is  not found, the program asks to insert a
USB storage device and it will look for the key file inside it.
If \fI<keyfile>\fR is "-" (dash), it will read stdin
.IP
When creating a tomb, this option can be used to specify the name (and
location) of the key you are creating. For example, you could use
.EX
tomb create -s 100 tombname -k /media/usb/tombname
.EE
to put the key on a usb pendrive

.B
.IP "-n"
Skip processing of post-hooks and bind-hooks if found inside the tomb.
See the \fIHOOKS\fR section in this manual for more information.
.B
.IP "-o"
Manually specify mount options to be used when opening a tomb instead
of the default \fIrw,noatime,nodev\fR. This option can be used to
mount a tomb read-only (ro) to prevent any modification of its data,
or to experiment with other settings (if you really know what you are
doing) see the mount(8) man page.
.B
.IP "--ignore-swap"
By default, Tomb will abort any create and open operation if swap is used (see
SWAP section for details). This flag will disable this behaviour. NOTE: it is
not secure to do so, unless you know that your swap is encrypted

.B
.IP "-h"
Display a help text and quit
.B
.IP "-v"
Display version and quit
.B
.IP "-q"
Run more quietly
.IP "-D"
Print more information while running, for debugging purposes


.SH HOOKS

Hooks are special files that can be placed inside the tomb and trigger
actions when it is opened and closed; there are two kinds of such
files: \fIbind-hooks\fR and \fIpost-hooks\fR can be placed in the
base root of the tomb.

.B
.IP "bind-hooks"
This hook file consists of a simple two column list of files or
directories inside the tomb to be made directly accessible inside the
current user's home directory. Tomb will use the "mount \-o bind"
command to bind locations inside the tomb to locations found in $HOME
so in the first column are indicated paths relative to the tomb and in
the second column are indicated paths relative to $HOME contents, for
example:

  mail          mail
  .gnupg        .gnupg
  .fmrc         .fetchmailrc
  .mozilla      .mozilla

.B
.IP "post-hooks"
This hook file gets executed as user by tomb right after opening it;
it can consist of a shell script of a binary executable that performs
batch operations every time a tomb is opened.

.SH PRIVILEGE ESCALATION

The tomb commandline tool needs to acquire super user rights to
execute most of its operations: to do so it uses sudo(8), while
pinentry(1) is adopted to collect passwords from the user.

Tomb executes as super user only those commands requiring it, while it
executes desktop applications as processes owned by the user.

.SH SWAP

During "create" and "open" operation, swap will complain and \fIabort\fR if
your system has swap activated. This can be annoying, and you can disable this
behaviour using \fI--ignore-swap\fR. Before doing that, however, you may be
interested in knowing the risks of doing so:
.IP \(bu
During both creation and opening it could write your secret key on the disk
.IP \(bu
After having opened the tomb, an application you're using could swap file
contents. So you'll put file contents in clear on your disk
.P

If you don't need swap, execute \fI swapoff -a\fR. If you really need it, you
could encrypt it. Tomb doesn't detect if your swap is encrypted, and will
complain anyway. In that case, using --ignore-swap is safe. Otherwise, use
--ignore-swap at your own risk



.SH BUGS
Please report bugs on the tracker at
.UR http://bugs.dyne.org
.UE

Get in touch with developers via mail using this
.UR http://dyne.org/contact
web page
.UE
or via chat on
.UR http://irc.dyne.org
.UE

.SH AUTHORS

Tomb is designed and written by Denis Roio aka Jaromil.

Tomb includes code by Anathema and Boyska.

Tomb's artwork is contributed by Jordi aka Mon Mort

Testing and reviews are contributed by Hellekin O. Wolf, Dreamer,
Shining, Mancausoft, Asbesto Molesto.

Cryptsetup is developed by Christophe Saout and Clemens Fruhwirth

.SH COPYING

This manual is Copyleft (c) 2011 Denis Roio <\fIjaromil@dyne.org\fR>

It includes contributions by Boyska

Permission is  granted to copy,  distribute and/or modify  this manual
under the terms of the  GNU Free Documentation License, Version 1.1 or
any  later   version  published  by  the   Free  Software  Foundation.
Permission is granted  to make and distribute verbatim  copies of this
manual page  provided the above  copyright notice and  this permission
notice are preserved on all copies.

.SH AVAILABILITY

The most recent version of Tomb sourcecode and up to date
documentation is available for download from its website on
\fIhttp://tomb.dyne.org\fR.

.SH SEE ALSO

.B
.IP cryptsetup(8)

GnuPG website on http://www.gnupg.org

DM-Crypt website on http://www.saout.de/misc/dm-crypt

LUKS website, http://code.google.com/p/cryptsetup


src/tomb-status.c
/*  Tomb - encrypted storage undertaker
 *  
 *  (c) Copyright 2007-2011 Denis Roio <jaromil@dyne.org>
 *
 * This source code is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Public License as published 
 * by the Free Software Foundation; either version 3 of the License,
 * or (at your option) any later version.
 *
 * This source code is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please refer to the GNU Public License for more details.
 *
 * You should have received a copy of the GNU Public License along with
 * this source code; if not, write to:
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <libgen.h>

#include <sys/types.h>
#include <sys/wait.h>

#include <gtk/gtk.h>
#include <libnotify/notify.h>

/* The Tomb icon is an artwork by Jordi aka MonMort
   a nomadic graffiti artist from Barcelona */
#include <monmort.xpm>

GdkPixbuf *pb_monmort;
GtkStatusIcon *status_tomb;
GtkMenu *menu_left, *menu_right;

NotifyNotification *notice;
GError *error;

char mapper[256];
char filename[256];
char mountpoint[256];

// forward declaration of callbacks
gboolean left_click(GtkWidget *w, GdkEvent *e);
gboolean cb_view(GtkWidget *w, GdkEvent *e);
gboolean cb_close(GtkWidget *w, GdkEvent *e);

gboolean right_click(GtkWidget *w, GdkEvent *e);
gboolean cb_about(GtkWidget *w, GdkEvent *e);


int main(int argc, char **argv) {
  GtkWidget *item_close, *item_view, *item_about;
  gint menu_x, menu_y;
  gboolean push_in = TRUE;

  char tomb_file[512];
  char tooltip[256];

  gtk_set_locale();
  gtk_init(&argc, &argv);

  // get the information from commandline
  if(argc<2) {
    fprintf(stderr, "error: need at least one argument, the path to a dm-crypt device mapper\n");
    exit(1);
  } else {
    // TODO: check if mapper really exists
    snprintf(mapper,255, "%s", argv[1]);
  }

  if(argc<3) sprintf(filename, "unknown");
  else snprintf(filename,255, "%s", argv[2]);

  if(argc<4) sprintf(mountpoint,"unknown");
  else snprintf(mountpoint,255, "%s", argv[3]);

  // libnotify
  notify_init(PACKAGE);

  // set and show the status icon
  pb_monmort = gdk_pixbuf_new_from_xpm_data(monmort);
  status_tomb = gtk_status_icon_new_from_pixbuf(pb_monmort);
  //  gtk_status_icon_set_name(status_tomb, "tomb");
  gtk_status_icon_set_title(status_tomb, "Tomb");

  snprintf(tooltip,255,"%s",mountpoint);
  gtk_status_icon_set_tooltip_text (status_tomb, tooltip);

  // LEFT click menu
  menu_left = (GtkMenu*) gtk_menu_new();
  // view
  item_view = gtk_menu_item_new_with_label("Explore");
  gtk_menu_attach(menu_left, item_view, 0, 1, 0, 1);
  g_signal_connect_swapped(item_view, "activate", G_CALLBACK(cb_view), NULL);
  gtk_widget_show(item_view);
  // close
  item_close = gtk_menu_item_new_with_label("Close");
  gtk_menu_attach(menu_left, item_close, 0, 1, 1, 2);
  g_signal_connect_swapped(item_close, "activate", G_CALLBACK(cb_close), NULL);
  gtk_widget_show(item_close);

  // connect it
  g_signal_connect_swapped(status_tomb, "activate", G_CALLBACK(left_click), menu_left);


  // RIGHT click menu
  menu_right = (GtkMenu*) gtk_menu_new();
  // about
  item_about = gtk_menu_item_new_with_label("About");
  gtk_menu_attach(menu_right, item_about, 0, 1, 0, 1);
  g_signal_connect_swapped(item_about, "activate", G_CALLBACK(cb_about), NULL);
  g_signal_connect_swapped(item_about, "popup-menu", G_CALLBACK(cb_about), NULL);
  gtk_widget_show(item_about);
  // connect it
  g_signal_connect_swapped(status_tomb, "popup-menu", G_CALLBACK(right_click), menu_right);

  // status icon
#if (HAVE_NOTIFY_NOTIFICATION_NEW_WITH_STATUS_ICON)
  notice = notify_notification_new_with_status_icon
    ("Tomb encrypted undertaker",
     "We started digging out bones",
     NULL, status_tomb);
#else
  notice = notify_notification_new
    ("Tomb encrypted undertaker",
     "We started digging out bones",
     NULL);
#endif
  notify_notification_set_icon_from_pixbuf(notice, pb_monmort);

  notify_notification_show(notice, &error);

  gtk_main();
  
  notify_uninit();

  exit(0);
  
}

// callbacks left click
gboolean left_click(GtkWidget *w, GdkEvent *e) {
  gtk_menu_popup(menu_left, NULL, NULL,
		 gtk_status_icon_position_menu, status_tomb,
		 1, gtk_get_current_event_time());
  return TRUE;
} 
gboolean cb_view(GtkWidget *w, GdkEvent *e) { 
  int pipefd[2];
  pid_t cpid;
  char buf;
  int c, res;
  char map[256];

  if (pipe(pipefd) <0) {
    fprintf(stderr,"pipe creation error: %s\n", strerror(errno));
    return FALSE;
  }

  cpid = fork();
  if (cpid == -1) {
    fprintf(stderr,"fork error: %s\n", strerror(errno));
    return FALSE;
  }
  if (cpid == 0) {    // Child
    close(pipefd[1]); // close unused write end
    for(c=0; read(pipefd[0], &buf, 1) > 0; c++)
      map[c] = buf;
    close(pipefd[0]);
    map[c] = 0;
    execlp("tomb-open", "tomb-open", map, (char*)NULL);
    _exit(1);
  }
  close(pipefd[0]); // close unused read end
  write(pipefd[1], mountpoint, strlen(mountpoint));
  close(pipefd[1]); // reader will see EOF

  return TRUE;
}


gboolean cb_close(GtkWidget *w, GdkEvent *e) { 
  int pipefd[2];
  pid_t cpid;
  char buf;
  int c, res;
  char map[256];

  if (pipe(pipefd) <0) {
    fprintf(stderr,"pipe creation error: %s\n", strerror(errno));
    return FALSE;
  }
  

  cpid = fork();
  if (cpid == -1) {
    fprintf(stderr,"fork error: %s\n", strerror(errno));
    return FALSE;
  }
  if (cpid == 0) {    // Child
    close(pipefd[1]); // close unused write end
    for(c=0; read(pipefd[0], &buf, 1) > 0; c++)
      map[c] = buf;
    close(pipefd[0]);
    map[c] = 0;
    execlp("tomb", "tomb", "close", map, (char*)NULL);
    _exit(1);
  }
  close(pipefd[0]); // close unused read end
  write(pipefd[1], mapper, strlen(mapper));
  close(pipefd[1]); // reader will see EOF

  waitpid(cpid, &res, 0);
  if(res==0) {
    gtk_main_quit();
    notify_uninit();
    exit(0);
  }
  /*  tomb-notify "Tomb '$tombname' is too busy."		\
      "Close all applications and file managers, then try again."
  */
  return TRUE;
}

// callbacks right click
gboolean right_click(GtkWidget *w, GdkEvent *e) {
  gtk_menu_popup(menu_right, NULL, NULL,
		 gtk_status_icon_position_menu, status_tomb,
		 1, gtk_get_current_event_time());
  return TRUE;
} 
gboolean cb_about(GtkWidget *w, GdkEvent *e) {
  const gchar *authors[] = {"Jaromil - http://jaromil.dyne.org",
			    "Code reviews and contributions by:",
			    "Dreamer, Hellekin O. Wolf, Shining, Mancausoft,",
			    "Anathema, Boyska and Nignux", NULL};
  const gchar *artists[] = {"Món Mort - http://monmort.blogspot.com",
			    "Asbesto Molesto - http://freaknet.org/asbesto",
			    NULL};
  GtkWidget *dialog = gtk_about_dialog_new();
  gtk_about_dialog_set_name(GTK_ABOUT_DIALOG(dialog), PACKAGE);
  gtk_about_dialog_set_version(GTK_ABOUT_DIALOG(dialog), VERSION); 
  gtk_about_dialog_set_copyright(GTK_ABOUT_DIALOG(dialog), 
				 "(C)2007-2011 Jaromil @ Dyne.org Foundation");
  gtk_about_dialog_set_artists(GTK_ABOUT_DIALOG(dialog), artists);
  gtk_about_dialog_set_authors(GTK_ABOUT_DIALOG(dialog), authors);

  gtk_about_dialog_set_comments(GTK_ABOUT_DIALOG(dialog), 
				"The Crypto Undertaker\n"
"\n"
"This program helps people keeping their bones together by taking care of their private data inside encrypted storage filesystems that are easy to access and transport.\n"
"\n"
"The level of security provided by this program is fairly strong: it uses  AES/SHA256 (cbc-essiv) to access the encrypted volumes in realtime so that all what is physically stored on your disc is only in an encrypted form.\n"
"Tomb also encourages users to separate keys from data, for instance storing them on USB.\n"
"\n"
);
  gtk_about_dialog_set_website(GTK_ABOUT_DIALOG(dialog), PACKAGE_URL);
  gtk_about_dialog_set_logo(GTK_ABOUT_DIALOG(dialog), pb_monmort);
  gtk_about_dialog_set_logo_icon_name(GTK_ABOUT_DIALOG(dialog), "monmort");
  // this below is active since gtk 3.0 so too early for it now
  //  gtk_about_dialog_set_license_type(GTK_ABOUT_DIALOG(dialog), GtkLicense.GTK_LICENSE_GPL_3_0);
  gtk_about_dialog_set_license(GTK_ABOUT_DIALOG(dialog),
"This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.\n"
"\n"
"This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n"
"\n"
"You should have received a copy of the GNU General Public License along with this program.\n"
"If not, see <http://www.gnu.org/licenses>\n"
"\n"
" The latest Tomb sourcecode is published on <http://tomb.dyne.org>\n");
  gtk_about_dialog_set_wrap_license(GTK_ABOUT_DIALOG(dialog), TRUE);
  gtk_dialog_run(GTK_DIALOG (dialog));
  gtk_widget_destroy(dialog);
  return TRUE;
}
  

  // GtkWidget *dialog = 
  //   gtk_message_dialog_new (NULL,
  // 			    GTK_DIALOG_DESTROY_WITH_PARENT,
  // 			    GTK_MESSAGE_INFO,
  // 			    GTK_BUTTONS_CLOSE,
  // 			    "Tomb '%s' open on '%s'\n"
  // 			    "device mapper: %s", filename, mountpoint, mapper);
  // gtk_dialog_run (GTK_DIALOG (dialog));
  // gtk_widget_destroy (dialog);
